src/multi.c

```diff
@@ -374,31 +374,32 @@ void touchWatchedKey(redisDb *db, robj *key) {
     }
 }
 
-/* On FLUSHDB or FLUSHALL all the watched keys that are present before the
- * flush but will be deleted as effect of the flushing operation should
- * be touched. "dbid" is the DB that's getting the flush. -1 if it is
- * a FLUSHALL operation (all the DBs flushed). */
-void touchWatchedKeysOnFlush(int dbid) {
-    listIter li1, li2;
+/* Set CLIENT_DIRTY_CAS to all clients of DB when DB is dirty.
+ * It may happen in the following situations:
+ * FLUSHDB, FLUSHALL, SWAPDB*/
+void touchWatchedKeysOnDirty(int dbid) {
+    listIter li;
     listNode *ln;
+    dictEntry *de;
 
-    /* For every client, check all the waited keys */
-    listRewind(server.clients,&li1);
-    while((ln = listNext(&li1))) {
-        client *c = listNodeValue(ln);
-        listRewind(c->watched_keys,&li2);
-        while((ln = listNext(&li2))) {
-            watchedKey *wk = listNodeValue(ln);
-
-            /* For every watched key matching the specified DB, if the
-             * key exists, mark the client as dirty, as the key will be
-             * removed. */
-            if (dbid == -1 || wk->db->id == dbid) {
-                if (dictFind(wk->db->dict, wk->key->ptr) != NULL)
-                    c->flags |= CLIENT_DIRTY_CAS;
+    redisDb *db = &server.db[dbid];
+    if (dictSize(db->watched_keys) == 0) return;
+
+    dictIterator *di = dictGetSafeIterator(db->watched_keys);
+    while((de = dictNext(di)) != NULL) {
+        robj *key = dictGetKey(de);
+        list *clients = dictGetVal(de);
+        if (!clients) continue;
+        listRewind(clients,&li);
+        while((ln = listNext(&li))) {
+            client *c = listNodeValue(ln);
+            if (dictFind(db->dict, key->ptr) != NULL) {
+                /* If the key exists, mark the client as dirty */
```

Comment:
this condition (if the key exists) is right on FLUSHDB, but on SWAPDB it probably needs to check both databases.
i.e. the WATCH should be invalidated if the key exists in either of them, and skipped only if it doesn't exist in either.

Reply:
> i.e. the WATCH should be invalidated if the key exists in either of them, and skipped only if it doesn't exist in either.

Thank you for your review, the current code, the following case (newly added, not yet submitted) failed, but you mean it should succeed, right?
```
test {SWAPDB is able to touch the watched keys} {
    r flushall
    r select 0
    r set x 30
    r watch x
    r swapdb 0 1
    r multi
    r ping
    r exec
} {}
```

We can solve this problem by advancing `touchWatchedKeysOnDirty` to before the dict exchange. The diff is as follows (comments have not been modified), WDYT?
```
diff --git a/src/db.c b/src/db.c
index 677ee2913..1b5b343a6 100644
--- a/src/db.c
+++ b/src/db.c
@@ -1321,6 +1321,9 @@ int dbSwapDatabases(long id1, long id2) {
     redisDb aux = server.db[id1];
     redisDb *db1 = &server.db[id1], *db2 = &server.db[id2];
 
+ touchWatchedKeysOnDirty(id1);
+ touchWatchedKeysOnDirty(id2);
+
     /* Swap hash tables. Note that we don't swap blocking_keys,
      * ready_keys and watched_keys, since we want clients to
      * remain in the same DB they were. */
@@ -1347,9 +1350,7 @@ int dbSwapDatabases(long id1, long id2) {
      * Also the swapdb should make transaction fail if there is any
      * client watching keys */
     scanDatabaseForReadyLists(db1);
-touchWatchedKeysOnDirty(id1);
     scanDatabaseForReadyLists(db2);
-touchWatchedKeysOnDirty(id2);
     return C_OK;
 }

```

Reply:
@yangbodong22011 but if we do that, then the other case will fail.

i.e. if i add a comment to your test:
```tcl
test {SWAPDB is able to touch the watched keys that exist} {
    r flushall
    r select 0
    r set x 30
    r watch x ;# make sure x (set to 30) doesn't change (SWAPDB will "delete" it)
    r swapdb 0 1
    r multi
    r ping
    r exec
} {}
```

but we can also test:
```tcl
test {SWAPDB is able to touch the watched keys that do not exist} {
    r flushall
    r select 1
    r set x 30
    r select 0
    r watch x ;# make sure the key x (currently missing) doesn't change (SWAPDB will create it)
    r swapdb 0 1
    r multi
    r ping
    r exec
} {}
```

Reply:
Yes, you are right. Let's return to your suggestion.

Reply:
Updated.